### ISSUE TITLE
Support generating sink Avro schemas with `timestamp-millis`

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#21559]
*Part of epic https://github.com/MaterializeInc/materialize/issues/17209.*

## Overview

Avro schemas support two types of timestamps: timestamps with millisecond precision and timestamps with microsecond precision. Today, Avro-formatted Kafka sinks always represent `timestamp`/`timestamptz` types in Avro as timestamps with microsecond precision, to match Materialize's internal representation of timestamps. But some users need their Avro timestamps to use millisecond precision, e.g. to meet their organization's internal policy for Avro schemas.

The goal here is to allow users to configure their Avro-formatted sinks to use timestamps with millisecond rather than microsecond precision.

### Option 1: new SQL type

Introduce a new type named `timestamptz_ms` which represents a timestamp with millisecond precision.

Internally, this type will wrap a `timestamptz`. It will support no operations (not even basic date/time operations) besides text input and output—not even binary input and output. Its sole purpose for now will be to serialize as `logicalType: timestamp-millis` in Avro schemas. 

In the future, if there is demand, we can add support for date/time operations to the type.

Given the following sink definition:

```sql
CREATE MATERIALIZED VIEW v AS SELECT '...'::timestamptz AS a, 'b'::timestamptz_ms AS b;
CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ... (
    OVERRIDE LOGICAL TYPE a = 'timestamp-millis', -- this is a bug, as the actual value is still in microseconds
    OVERRIDE LOGICAL TYPE b = 'timestamp-millis', -- this is ok, because the materialized view properly converts to milliseconds
);
```

Materialize would generate the following Avro schema:

```json
{
  "type": "record",
  "name": "row",
  "fields" : [
    {"name": "a", "type": "long", "logicalType": "timestamp-micros"},
    {"name": "b", "type": "long", "logicalType": "timestamp-millis"}
  ]
}
```

Pros:

  * Can vary precision on a field-by-field basis
  * Hard, if not impossible, to misuse
  * `timestamptz_ms` will be useful for other data types in the future that have variable precision for timestamps
  * Fits with philosophy of matching semantics between Materialize and Avro

Cons:

  * Implementing a new SQL type is a medium lift
  * The lack of date/time operations on the `timestamptz_ms` type may be confusing

### Option 2: `LOGICAL TYPE` override

Given the following sink definition:

```sql
CREATE MATERIALIZED VIEW v AS SELECT '...'::timestamptz AS a, extract(epoch from '...'::timestamptz)/1000 AS b, 'c'::timestamptz AS c;
CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ... (
    OVERRIDE LOGICAL TYPE a = 'timestamp-millis', -- this is a bug, as the actual value is still in microseconds
    OVERRIDE LOGICAL TYPE b = 'timestamp-millis', -- this is ok, because the materialized view properly converts to milliseconds
);
```

Materialize would generate the following Avro schema:

```json
{
  "type": "record",
  "name": "row",
  "fields" : [
    {"name": "a", "type": "long", "logicalType": "timestamp-millis"},
    {"name": "b", "type": "long", "logicalType": "timestamp-millis"},
    {"name": "c", "type": "long", "logicalType": "timestamp-micros"}
  ]
}
```

Pros: 

 * Can vary precision on a field-by-field basis
 * Permits configuring of arbitrary logical types (though it's not clear there are other logical types that need this kind of configuration; all other logical types in the Avro spec has a 1:1 mapping with Materialize types)

Cons:

 * Easy to misuse, e.g., by failing to actually convert timestamp to milliseconds

### Option 3: per-sink override

In this option, we allow configuring a per-sink `TIMESTAMP PRECISION` option.

Given the following sink definition:

```sql
CREATE MATERIALIZED VIEW v AS SELECT '...'::timestamptz AS a, 'b'::timestamptz AS b;
CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ... (
    TIMESTAMP PRECISION = 'milliseconds' -- the default is "microseconds"
);
```

Materialize would generate the following Avro schema:

```json
{
  "type": "record",
  "name": "row",
  "fields" : [
    {"name": "a", "type": "long", "logicalType": "timestamp-millis"},
    {"name": "b", "type": "long", "logicalType": "timestamp-millis"}
  ]
}
```

Note the option is part of the `FORMAT AVRO` clause and not part of the overall `WITH` options for the sink.

Pros:

  * Obvious syntax
  * Difficult, if not impossible, to misuse

Cons:

  * Cannot specify timestamp precision on a field by field basis

### Option 4: per-field override

This is like option 3, but the `TIMESTAMP PRECISION` option is specified per field rather than per sink.

Given the following sink definition:

```sql
CREATE MATERIALIZED VIEW v AS SELECT '...'::timestamptz AS a, 'b'::timestamptz AS b;
CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ... (
    TIMESTAMP PRECISION ON a = 'milliseconds' -- the default is "microseconds"
);
```

Materialize would generate the following Avro schema:

```json
{
  "type": "record",
  "name": "row",
  "fields" : [
    {"name": "a", "type": "long", "logicalType": "timestamp-millis"},
    {"name": "b", "type": "long", "logicalType": "timestamp-micros"}
  ]
}
```

Note the option is part of the `FORMAT AVRO` clause and not part of the overall `WITH` options for the sink.

Pros:

  * Obvious syntax
  * Difficult, if not impossible, to misuse

Cons:

  * We don't know how to do general-purpose field selectors (`ON a.b.c`).
  * Not in line with philosophy of matching semantics between Materialize and Avro as much as possible

See also the cons on the alternative for including comments in Avro sink schemas (https://github.com/MaterializeInc/materialize/issues/21557), which is quite similar to this option.

## Next steps

The next step is to align on a design. If we can't get consensus on the best option via a Slack conversation, we may need to promote this to a formal design document and run through that process.

cc @dseisun-materialize @moulimukherjee @bkirwi @umanwizard @guswynn @sjwiesman

**Comment by umanwizard** [comment]
Thanks for the detailed writeup!

I would be fine with either 1 or 4. I think 2 is way too big of a footgun, and I don't like the fact that 3 forces you to make the decision once for the entire schema.

In isolation I prefer 1 because it's indeed weird to be able to narrow the precision of one very specific kind of thing in this way but not, e.g., narrow doubles to floats in a similar way. If we go with the SQL-type-based solution, then everything works in a consistent way.

However, if 4 is way easier to implement, I'm also fine with that.

This "con" for 4:

> We don't know how to do general-purpose field selectors (ON a.b.c).

is not quite true. I invented such a way for specifying distributions in kgen, which is encoded in the [`RandomAvroGenerator::new_inner`](https://github.com/MaterializeInc/materialize/blob/2d71ae0581/src/kafka-util/src/bin/kgen.rs#L295-L300) function. I just never documented it.

I'll attempt to document and formalize it here.

An Avro schema node  can appear in the following contexts:

1. The top-level node of the schema,
2. One of the fields of a record,
3. One of the branches of a union,
4. The value type of an array, or
5. The value type of a map.

We define the "unambiguous name" of a schema node recursively as follows:

* If the node is of a [named type](https://avro.apache.org/docs/1.10.2/spec.html#names) (record, enum, or fixed) then its unambiguous name is the same as its "fullname" as defined by the Avro spec; e.g. just its name if it is not part of a namespace, or its namespace, followed by `.`, followed by its name if it is.
* Otherwise, if the node is the top-level node of the schema, its unambiguous name is the empty string (we will never have to deal with this, as sinks in Materialize can't have a schema that is just a timestamp field and nothing else. Things are always at least nested inside a record.)
* Otherwise, if the node is a record field, its unambiguous name is the record's unambiguous name (i.e., its fullname), followed by `::`, followed by the field name. Note that you don't need to account for multiple levels of nesting here as you sort of imply with your `ON a.b.c` example -- if record `my.namespace.a` contains record `my.namespace.b` which contains record `my.namespace.c` which contains field `f1`, you can just call that field `my.namespace.c::f1`, unambiguously.
* Otherwise, if the node is one of the variants of a union, its unambiguous name is the union's unambiguous name, followed by a `.`, followed by the node's index within a union. I think we should actually change this and use the type name instead (e.g. `union_unambiguous_name.long` instead of `union_unambiguous_name.4`). That would work because unions are not allowed to contain multiple variants of the same kind of type unless they are named types.
* Otherwise, if the node is the value type of an array or map, its unambiguous name is the array or map's unambiguous name, followed by `[]`.

**Comment by umanwizard** [comment]
to cut through the details a bit here: 99% of the time in the real world, people will be referring to fields in records, which will look like `my.namespace.my_record::my_field`.

**Comment by benesch** [comment]
I just remembered that PostgreSQL allows specifying the precision of a timestamp explicitly!

> time, timestamp, and interval accept an optional precision value p which specifies the number of fractional digits retained in the seconds field. By default, there is no explicit bound on precision. The allowed range of p is from 0 to 6.

—https://www.postgresql.org/docs/current/datatype-datetime.html

So, "problem solved." If a timestamp's precision is <= 3, we should just emit `timestamp-millis` instead of `timestamp-micros`.

**Comment by benesch** [comment]
I figured the design was sufficiently obvious in light of these timestamp precisions that I put up a ready-for-review PR with that approach: #21570.

**Comment by moulimukherjee** [comment]
This looks related as well https://github.com/MaterializeInc/materialize/issues/10837

**Comment by benesch** [comment]
Oh wow, indeed. I'll add "Fixes #10837" to the timestamp precision PR.
